mjolnir/core/GBAOABLangevinIntegrator.hpp

```diff
@@ -0,0 +1,338 @@
+          r_dt_in_correction_(1 / dt_in_correction_),
+          old_position_(gammas_.size()),
+          old_pos_rattle_(gammas_.size())
+    {}
+    ~GBAOABLangevinIntegrator() = default;
+
+    void initialize(system_type& sys, forcefield_type& ff, rng_type& rng);
+
+    real_type step(const real_type time, system_type& sys, forcefield_type& ff,
+                   rng_type& rng);
+
+    void update(const system_type& sys)
+    {
+        if(!sys.has_attribute("temperature"))
+        {
+            throw std::out_of_range("mjolnir::g-BAOABLangevinIntegrator: "
+                "Langevin Integrator requires reference temperature, but "
+                "`temperature` is not found in `system.attribute`.");
+        }
+        this->temperature_ = sys.attribute("temperature");
+        this->reset_parameter(sys);
+        return;
+    }
+
+    real_type delta_t() const noexcept {return dt_;}
+    std::vector<real_type> const& parameters() const noexcept {return gammas_;}
+
+  private:
+
+    void reset_parameter(const system_type& sys) noexcept
+    {
+        const auto kBT = physics::constants<real_type>::kB() * this->temperature_;
+        for(std::size_t i=0; i<sys.size(); ++i)
+        {
+            const auto gamma     = this->gammas_.at(i);
+            const auto gamma_dt  = -1 * gamma * this->dt_;
+            this->exp_gamma_dt_.at(i) = std::exp(gamma_dt);
+            this->noise_coeff_ .at(i) = std::sqrt(
+                    kBT * (1 - std::exp(2 * gamma_dt)) * sys.rmass(i));
+        }
+        return;
+    };
+
+    coordinate_type gen_R(rng_type& rng) noexcept
+    {
+        const auto x = rng.gaussian();
+        const auto y = rng.gaussian();
+        const auto z = rng.gaussian();
+        return math::make_coordinate<coordinate_type>(x, y, z);
+    }
+
+    void correct_coordinate(system_type& sys, const real_type tolerance)
+    {
+        MJOLNIR_GET_DEFAULT_LOGGER();
+
+        std::size_t rattle_step = 0;
+        while(rattle_step<correction_max_iter_)
+        {
+            bool corrected = false;
+            for(std::size_t i=0; i<this->constraints_.size(); ++i)
+            {
+                const auto& constraint = constraints_[i];
+                const auto& indices    = constraint.first;
+                auto& p1  = sys.position(indices[0]);
+                auto& p2  = sys.position(indices[1]);
+                auto& v1  = sys.velocity(indices[0]);
+                auto& v2  = sys.velocity(indices[1]);
+                auto& op1 = this->old_pos_rattle_[indices[0]];
+                auto& op2 = this->old_pos_rattle_[indices[1]];
+                auto& rm1 = sys.rmass(indices[0]);
+                auto& rm2 = sys.rmass(indices[1]);
+
+                const auto      dp = p2 - p1;
+                const real_type dp2 = math::length_sq(dp);
+                const real_type missmatch = square_v0s_[i] - dp2;
+
+                if(abs(missmatch) > tolerance)
```

Comment:
Don't forget `std::` here.
Also, mjolnir prefer `<` than `>` because we consider it is easier to understand if it is unified. Could you use `<` whenever possible?

```suggestion
                if(tolerance < std::abs(missmatch))
```